torch/distributions/transforms.py

```diff
@@ -535,6 +536,83 @@ def log_abs_det_jacobian(self, x, y):
         return result.expand(shape)
 
 
+class CorrCholeskyTransform(Transform):
+    r"""
+    Transforms an uncontrained real vector :math:`x` with length :math:`D*(D-1)/2` into the
+    Cholesky factor of a D-dimension correlation matrix. This Cholesky factor is a lower
+    triangular matrix with positive diagonals and unit Euclidean norm for each row.
+    The transform is processed as follows:
+
+        1. First we convert x into a lower triangular matrix in row order.
+        2. For each row :math:`X_i` of the lower triangular part, we apply a *signed* version of
+        class :class:`StickBreakingTransform` to transform :math:`X_i` into a
+        unit Euclidean length vector using the following steps:
+            a. Scales into the interval :math:`(-1, 1)` domain: :math:`r_i = \tanh(X_i)`.
+            b. Transforms into an unsigned domain: :math:`z_i = r_i^2`.
+            c. Applies :math:`s_i = StickBreakingTransform(z_i)`.
+            d. Transforms back into signed domain: :math:`y_i = sign(r_i) * \sqrt{s_i}`.
+    """
+    domain = constraints.real_vector
+    codomain = constraints.corr_cholesky
+    bijective = True
+    # Note that since we add a dim in _call this should actually be 2 for the inverse transform.
+    event_dim = 1
```

Comment:
@fehiepsi, @fritzo - Note that the forward transform add an additional dim, so `event_dim=1` is defined w.r.t. to the value passed to the forward transform. The inverse transform's event_dim is consequently incorrect. I wasn't sure what to do. One thing would be to modify the inverse transform for this to have `event_dim=2`. Any other suggestions?

Reply:
I am not sure what we can do here... Either leaving this `1` or `2` is fine I guess.

Reply:
Good point, I believe we did not anticipate transforms that have change event dim? How about we add new properties to the base `Transform` class and then override these in your new transform?

<details>

```py
class Transform:
    @property
    def input_event_dim(self):
        return self.event_dim

    @property
    def output_event_dim(self):
        return self.event_dim

class _InverseTransform(Transform):
    @property
    def input_event_dim(self):
        return self._inv.output_event_dim

    @property
    def output_event_dim(self):
        return self._inv.input_event_dim

class CorrCholeskyTransform(Transform):
    input_event_dim = 1
    output_event_dim = 2

    @property
    def event_dim(self):
        raise ValueError("Please use .intput_event_dim or .output_event_dim instead")
```

</details>

Reply:
That sounds like a good idea. I'll make the change!

Reply:
Done. Thanks for the suggestion!